src/Functions/FunctionsExternalDictionaries.h

```diff
@@ -594,31 +615,122 @@ class FunctionDictGetNoType final : public IFunction
 
 private:
 
+    std::pair<ColumnPtr, ColumnPtr> getDefaultsShortCircuit(
+        IColumn::Filter & default_mask,
+        const DataTypePtr & result_type,
+        const ColumnWithTypeAndName & last_argument) const
+    {
+        auto rows = default_mask.size();
+
+        auto mask_col = ColumnUInt8::create();
+        mask_col->getData() = std::move(default_mask);
+        ColumnPtr mask_col_res = std::move(mask_col);
+
+        IColumn::Filter mask(rows, 1);
+        auto mask_info = extractInvertedMask(mask, mask_col_res);
+        ColumnWithTypeAndName column_before_cast = last_argument;
+        maskedExecute(column_before_cast, mask, mask_info);
+
+        default_mask = std::move(mask);
+
+        ColumnWithTypeAndName column_to_cast = {
+            column_before_cast.column->convertToFullColumnIfConst(),
+            column_before_cast.type,
+            column_before_cast.name};
+        auto casted = IColumn::mutate(castColumnAccurate(column_to_cast, result_type));
+        casted->expand(default_mask, false);
+        return {std::move(casted), mask_col_res};
+    }
+
+    void restoreShortCircuitColumn(
+        ColumnPtr & result_column,
+        ColumnPtr defaults_column,
+        ColumnPtr mask_column,
+        IColumn::Filter & inverted_mask,
+        const DataTypePtr & result_type) const
+    {
+        auto mut_lhs = IColumn::mutate(std::move(result_column));
+        mut_lhs->expand(inverted_mask, true);
+
+        auto if_func = FunctionFactory::instance().get("if", helper.getContext());
+        ColumnsWithTypeAndName if_args =
+        {
+            {mask_column, std::make_shared<DataTypeUInt8>(), {}},
+            {std::move(mut_lhs), result_type, {}},
+            {defaults_column, result_type, {}},
+        };
+
+        auto rows = inverted_mask.size();
+        result_column = if_func->build(if_args)->execute(if_args, result_type, rows);
+    }
```

Comment:
So, I see that right now we have the following logic:
After calling `getColumnsOrDefaultShortCircuit` we got short result columns and `default_mask` that contains `0` for rows where we should place defaults. Then we execute our lazy default function using inverted `default_mask` and get short columns with default values. After it we expand both short columns and use `if(reversed_default_mask, expanded_result_column, expadefault_column)`

But actually we can avoid expanding columns at all, because `if` function should handle short columns as it supports short circuit arguments execution (in `if` in `executeImpl` we execute lazy arguments using mask and then just work with 2 short columns). Also, if we will return mask with `1` on default values from `getColumnsOrDefaultShortCircuit` we won't need to inverse it

Reply:
It would be great to avoid expand and inverse, which took me much time to understand. Let me give it a try.

Reply:
It's great to see the logic becomes easier after removing the `expand` function. But I think the `inverse` would happen at least once even with 1 and 0 switched, because we need the cond of if statement in `restoreShortCircuitColumn`, while the cond column directly use `default_mask` actually. This also explains the reason 1 represents the found value while 0 means default (think of true and false in `if then else` statement).

Reply:
> But I think the inverse would happen at least once even with 1 and 0 switched, because we need the cond of if statement in restoreShortCircuitColumn, while the cond column directly use default_mask actually

But can we just switch then/else statement in `if` function? Like, `if (default_mask, default_column, result_column)` and no need to inverse anything

Reply:
switched and removed the inverse.